[Reader Manuals Consolidation] Display only supported readers

iamgabrielma opened the pull request:
Closes: #7219 

### Description
At the moment we're showing all Manuals in the consolidated view despite not all readers being available to all regions.

This PR modifies the way we initialize `CardReaderManualsViewModel` by checking the supported readers for each region, so we only display the manuals for the supported readers based on the Store country, that's:

- US-based stores will only display the manuals for Chipper and M2.
- CA-based stores will only display the manual for Wisepad 3.


### Testing instructions
- In your US or CA store, go to Settings > In-Person Payments > Card Reader manuals:
-- See manuals for Chipper and M2 in the US store.
-- See the manual for Wisepad 3 in the CA store.

### Screenshots

![RMC 2022-07-11 filter by country](https://user-images.githubusercontent.com/3812076/178209199-5b062ba8-f170-416f-87b6-33576fd4da64.gif)


---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

iamgabrielma committed:
Initialize viewmodel only with supported readers

iamgabrielma committed:
Add Unit Tests CardReaderManualsViewModel

iamgabrielma committed:
Add foreground and background colors to the View

iamgabrielma committed:
Unit Tests for CardReaderManualsViewModel

joshheald commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardReaderManualsViewModel.swift:
```diff
@@ -2,17 +2,21 @@ import Foundation
+    var configurationLoader: CardPresentConfigurationLoader
     let manuals: [Manual]
 
     init(cardReaders: [CardReaderType] = [.chipper, .stripeM2, .wisepad3]) {
```
The `cardReaders` parameter is unused now. It might be worth passing in the configuration for testability here, but you can definitely remove `cardReaders`.

iamgabrielma replied:
That sounds like a good idea, for the moment I'll remove the `cardReaders` and explore passing the config on a different [enhancement issue](https://github.com/woocommerce/woocommerce-ios/issues/7275) 👍

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/CardPresentPayments/CardReaderManualsViewModelTests.swift:
```diff
@@ -0,0 +1,107 @@
+
+class CardReaderManualsViewModelTests: XCTestCase {
+
+    private var storageManager: MockStorageManager!
+    private var stores: MockStoresManager!
+    private let sampleSiteID: Int64 = 1234
+
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+        storageManager = MockStorageManager()
+        stores = MockStoresManager(sessionManager: .makeForTesting(authenticated: true))
+        stores.sessionManager.setStoreId(sampleSiteID)
+        ServiceLocator.setSelectedSiteSettings(SelectedSiteSettings(stores: stores, storageManager: storageManager))
+    }
+
+    override func tearDownWithError() throws {
+        ServiceLocator.setSelectedSiteSettings(SelectedSiteSettings())
+        storageManager.reset()
+        storageManager = nil
+        stores = nil
+        try super.tearDownWithError()
+    }
+
+    func test_viewModel_not_nil() {
+        // Given
+        let viewModel = CardReaderManualsViewModel()
+
+        // Then
+        XCTAssertNotNil(viewModel)
+    }
```
Well done for having a good go at unit testing!

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/CardPresentPayments/CardReaderManualsViewModelTests.swift:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    func test_viewModel_when_init_then_has_default_manuals() {
+         // Given
+         let viewModel = CardReaderManualsViewModel()
+
+         // When
+         let expectedManuals = viewModel.manuals
+
+         // Then
+         XCTAssertEqual(viewModel.manuals, expectedManuals)
```
```suggestion
         // Given
        let viewModel = CardReaderManualsViewModel()
        let expectedManuals = [CardReaderType.chipper.manual, CardReaderType.stripeM2.manual]

        // When
        let actualManuals = viewModel.manuals

        // Then
        assertEqual(expectedManuals, actualManuals)
```

Slight changes here:
1. This defines the expectedManuals in a way that you can read what they are, and avoids later comparing that `viewModel.manuals == viewModel.manuals`, which will always be true.
2. We put `expectedManuals` in the Given section: it's an axiom, not an action
3. We're testing `viewModel.manuals`, so we put that in the When section
4. We can use the `assertEqual` helper, rather than `XCTAssertEqual`, which will tell us what the differences are between the two when the test fails.

joshheald replied:
When I run this test, I get the following:
<img width="960" alt="CleanShot 2022-07-11 at 15 48 12@2x" src="https://user-images.githubusercontent.com/2472348/178292150-16683c5e-c0cd-4125-9859-fa3fddad73ff.png">

> test failure: CardReaderManualsViewModelTests.test_viewModel_when_init_then_has_default_manuals() failed: assertEqual failed: Found difference for 
> Different count:
> |	Received: (0) []
> |	Expected: (2) [ ... ]

This will have passed for you because the `viewModel.manuals` returned 0 elements in both cases.

```
po configurationLoader.configuration
▿ CardPresentPaymentsConfiguration
  - countryCode : ""
  - paymentMethods : 0 elements
  - currencies : 0 elements
  - paymentGateways : 0 elements
  - supportedReaders : 0 elements
```

The fix is to set a configuration, which you can pass in to `CardReaderManualsViewModel` in the init method, and allow it to use the default configuration at runtime, which we know works from our manual testing. e.g.:

```
// Given
let viewModel = CardReaderManualsViewModel(cardPresentConfiguration: CardPresentPaymentsConfiguration(country: "US"))
```

iamgabrielma replied:
Thanks! This is very helpful!

> avoids later comparing that viewModel.manuals == viewModel.manuals, which will always be true.

Oops, this one went unnoticed 😅 For the moment I'll remove the faulty test as we're already testing the 3 possibilities in the other tests and continue with César to see how can we improve these 👍

joshheald reviewed (approved):
Works well in general 👍

### Tests
Nice one for doing the unit tests, and finding your way through test data setup.

I've gone through one test in detail, which isn't testing what you want it to at the moment and can be rewritten to be a bit easier to read and more expressive – I hope that's helpful, but feel free to follow up with César or Jirka for their views/help with any questions about this approach.

### Behaviour
At first I thought it wasn't consistently opening the web view when I tapped the manual rows, but I discovered that was because it was only noticing taps directly on the image, text, or disclosure indicator, not on the space between them. I suspect this is from previous work on the manuals, so sorry I didn't spot it previously, but it's definitely worth fixing as the next thing. 

No need to do it in this PR, but see if you can figure it out next. It may be a problem with the shared component we used, in which case make sure you understand any wider implications of the fix as you change it.

iamgabrielma commented:
> At first I thought it wasn't consistently opening the web view when I tapped the manual rows, but I discovered that was because it was only noticing taps directly on the image, text, or disclosure indicator, not on the space between them. I suspect this is from previous work on the manuals, so sorry I didn't spot it previously, but it's definitely worth fixing as the next thing.
> 
> No need to do it in this PR, but see if you can figure it out next. It may be a problem with the shared component we used, in which case make sure you understand any wider implications of the fix as you change it.

I see what you mean, I didn't realize it either during the testing. Opened a new issue here: https://github.com/woocommerce/woocommerce-ios/issues/7274

iamgabrielma committed:
Remove CardReaderType init param and defaultvalues

iamgabrielma committed:
Remove faulty unit test

At the moment this test is merely checking viewModel.manuals == viewModel.manuals, a different approach for testing the viewModel on init could be to pass a configuration parameter.
